[+] Support for floating point signals.

Ezhik452 opened the pull request:
Hi everyone!

This PR introduces support of floating point signals to golang code generation from DBC files. Floating point signals differs in the next aspects:
* Values are stored using bit representation of the floating point.
* Because of the previous point conversion needs to be a reinterpret cast on top if the bit representation. For example, for 32-bit float it will be `(*float32)(unsafe.Pointer(&i))`
* Float nature of the value is represented in DBC file using `SIG_VALTYPE_ <message_id> <signal_name> : <type_val>;`. `type_val` can be `0` for integer values, `1` for float32 and `2` for float64. Parsing of this is implemented in library already.

This PR doesn't touch next topics for now:
* Testing of the generated files. If tests are needed I'll be grateful if you point me on how tests have to be implemented for this library.
* Signed/Unsigned indication (`1+`, `1-`) for the float values. For me it looks like this indication doesn't make difference for floats, but I lack documentation of this part and devices that I can test it with only use signed values (`1-`).
* Implementation for float64 is extrapolation of my knowledge and tests form the float32 implementation. Feel free to correct me if you have more insight.
* Implementation uses signal length to deduct signal type (float32 vs float64) and do not verify that it complies with value in `SIG_VALTYPE_`. Maybe some checks need to be added.

P.S.: It's my first PR to the open-source, so don't be too harsh, please :)

---
Example DBC file:
```
BO_ 42 Float32Message: 8 Vector__XXX
	SG_ Float32ValueNoRange : 0|32@1- (1,0) [0|0] "" Vector__XXX
	SG_ Float32WithRange : 32|32@1- (1,0) [-100|100] "" Vector__XXX

SIG_VALTYPE_ 42 Float32ValueNoRange: 1;
SIG_VALTYPE_ 42 Float32WithRange: 1;

BO_ 43 Float64Message: 8 Vector__XXX
	SG_ Float64ValueNoRange : 0|64@1- (1,0) [0|0] "" Vector__XXX

SIG_VALTYPE_ 43 Float64ValueNoRange: 2;

BO_ 44 AnotherFloat64Message: 8 Vector__XXX
	SG_ Float64ValueWithRange : 0|64@1- (1,0) [-100|100] "" Vector__XXX

SIG_VALTYPE_ 44 Float64ValueWithRange: 2;

```

Jassob commented:
Hi @Ezhik452 and thank you for your contribution!

This feature is not something we use internally at Einride, so we might be a bit slow to review your proposal. 

Personally I also need to review/refresh my knowledge of DBC files again.

Is the `SIG_VALTYPE_`part of nominal DBC files?

Ezhik452 commented:
Hi @Jassob 
> This feature is not something we use internally at Einride, so we might be a bit slow to review your proposal.

No problem, thanks for taking time :)

> Is the `SIG_VALTYPE_`part of nominal DBC files?

I believe so. I don't have full standard documentation, but there is a description in paragraph 8.2  (pages 5-7) of the [DBC format description](http://www.mcu.so/Microcontroller/Automotive/dbc-file-format-documentation_compress.pdf)

Ezhik452 commented:
I removed `float64` signals and modified parse and compile tests to accommodate to changes in the `example.dbc`.

![Screenshot_20240517_152251](https://github.com/einride/can-go/assets/64531926/ab8fbf7d-eb65-418e-8f18-40d23af22045)

Jassob commented:
@Ezhik452 Perfect, thank you! 

Would it be possible for you to merge the commits into one? Currently you have the first commit that adds the floating points types and then some fixups to unify float32 and float64 and then we remove float64 again in later commits.

It will take me much less time to look through your code and understand your changes if I don't have to skip through changes that gets fixed in later commits (I review commit per commit).

Jassob commented on testdata/dbc/example/example.dbc:
```diff
@@ -39,6 +39,10 @@ BO_ 500 IODebug: 6 IO
  SG_ TestScaledEnum : 40|2@1+ (2,0) [0|6] "" DBG
 
+BO_ 600 IOFloat32: 8 IO
+	SG_ Float32ValueNoRange : 0|32@1- (1,0) [0|0] "" DBG
+	SG_ Float32WithRange : 32|32@1- (1,0) [-100|100] "" DBG
```
Please follow the same kind of indentation as the rest of the DBC file (i.e. with one space indentation and not one tab). The difference is visible in editors with different tab stops :)

```suggestion
 SG_ Float32ValueNoRange : 0|32@1- (1,0) [0|0] "" DBG
 SG_ Float32WithRange : 32|32@1- (1,0) [-100|100] "" DBG
```

Jassob replied:
Otherwise great test additions! :raised_hands:

Ezhik452 replied:
Can you, please, point me to example of the negative test in the repo? I will add some similar tests.

Jassob replied:
I'm afraid there aren't any yet, but it's basically what `internal/generate/compile_test.go`does, but with invalid DBC snippets

Ezhik452 replied:
Thanks, I will take a look.

Ezhik452 replied:
Added a couple negative tests.

Jassob replied:
Great, thank you!

Jassob commented on internal/generate/compile.go:
```diff
@@ -98,6 +98,21 @@ func (c *compiler) collectDescriptors() {
+				c.addWarning(&compileError{def: def, reason: "no declared signal"})
+				continue
+			}
+			err := def.SignalValueType.Validate()
+			if err != nil {
+				c.addWarning(&compileError{def: def, reason: err.Error()})
+			}
+			if def.SignalValueType == dbc.SignalValueTypeFloat32 {
+				signal.IsFloat = true
+			} else {
+				signal.IsFloat = false
+			}
```
I think we should only allow values of 0 and 1, so we could actually inline the `def.SignalValueType.Validate()` here.

```suggestion
			switch def.SignalValueType {
			case dbc.SignalValueTypeInt:
				signal.IsFloat = false
			case dbc.SignalValueTypeFloat32:
				signal.IsFloat = true
			default:
				reason := fmt.Sprintf("unsupported signal value type: %v", def.SignalValueType)
				c.addWarning(&compileError{def: def, reason: reason})
			}
```

What do you think about that?

Ezhik452 replied:
Looks good to me, thanks!

Jassob commented on testdata/dbc/example/example.dbc:
```diff
@@ -77,3 +81,6 @@ VAL_ 100 Command 3 "Headlights On" 2 "Reboot" 1 "Sync" 0 "None" ;
 VAL_ 500 TestEnum 2 "Two" 1 "One" ;
 VAL_ 500 TestScaledEnum 3 "Six" 2 "Four" 1 "Two" 0 "Zero" ;
 VAL_ 500 TestBoolEnum 1 "One" 0 "Zero" ;
+
+SIG_VALTYPE_ 600 Float32ValueNoRange: 1;
+SIG_VALTYPE_ 600 Float32WithRange: 1;
```
:ok_hand:

Jassob reviewed (commented):
I think this looks really good, thank you for your persistence in working with us here! :clap: 

I would be very happy if you could also add some negative tests in the testdata folder, i.e. some tests that should actually fail to compile with our current changes. Test cases that come to mind are for example having a float64 signal, or having an invalid SIG_VALTYPE_ declaration pointing to a non-existing signal.

Ezhik452 commented:
> @Ezhik452 Perfect, thank you!
> 
> Would it be possible for you to merge the commits into one? Currently you have the first commit that adds the floating points types and then some fixups to unify float32 and float64 and then we remove float64 again in later commits.
> 
> It will take me much less time to look through your code and understand your changes if I don't have to skip through changes that gets fixed in later commits (I review commit per commit).

No problem, I will do it.

Jassob reviewed (commented):
Starting to look really good and thank you for the negative test cases!

I'm still missing test cases for checking if the bit pattern on the CAN frames are what we expected and if we can read/write messages that are compatible with other tools.

Do you have the possibility to get access to some dumped messages from a node that outputs signals with float types? If so, it would be great to check that we can properly decode such a message.

Otherwise, a good first test would be to make sure that the following property holds: 
```
if err != m2.UnmarshalFrame(m1.Frame()); err != nil {
  t.Fatalf("Failed to unmarshal frame: %v", err)
}
if !reflect.DeepEqual(m1, m2) {
  t.Fatalf("Frames are not equal: %v != %v", m1, m2)
} 
```

For different values of the float signals. Also please add a test that checks that we can set Float32ValueNoRange and Float32WithRange signals in `internal/generate/example_test.go` (which is also where I'd add the marshal/unmarshal test).

Finally, please also reword the commit to conform with https://www.conventionalcommits.org/en/v1.0.0/ (i.e. call it something like 

> feat: add support for floating point signals
> 
> This commit adds support for the SIG_VALTYPE_ DBC declaration for 32-bit floating point signals, in addition to the already existing support for integer typed signals.

I need to add a CONTRIBUTING.md to this repository, so you couldn't have known and I'm sorry for causing you extra work. Please let me know if you want any help :)

Ezhik452 commented on internal/generate/compile_test.go:
```diff
@@ -306,3 +334,31 @@ func TestCompile_ExampleDBC(t *testing.T) {
+	if err != nil {
+		t.Fatal(err)
+	}
+	// We expect one warning for incorect signal name in SIGVAL_TYPE_ declaration
```
Fixed, thanks!

Ezhik452 commented on internal/generate/file.go:
```diff
@@ -841,6 +843,8 @@ func signalType(m *descriptor.Message, s *descriptor.Signal) string {
 	var t types.BasicKind
 	switch {
+	case s.Length <= 32 && s.IsFloat:
+		t = types.Float32
```
To think of it, it's invalid case, probably. Reinterpret cast for non 32-bit value to 32-bit float will not work properly.
I changed condition to equal and added warning for it in DBC parsing in `compile.go`. Also I added negative test for incorrect signal length for float32 signals.

Ezhik452 commented:
I added float32 tests to Marshal/Unmarshal tests in `internal/generate/example_test.go`. This tests revealed Marshal bug related to the size of reinterpreted float value. It's fixed in `signal.go` on line 186.
 
I don't fully understands what tests are required for:
> I'm still missing test cases for checking if the bit pattern on the CAN frames are what we expected and if we can read/write messages that are compatible with other tools.

Commit is reworded according to guideline.

Jassob commented:
> I added float32 tests to Marshal/Unmarshal tests in `internal/generate/example_test.go`. This tests revealed Marshal bug related to the size of reinterpreted float value. It's fixed in `signal.go` on line 186.

Awesome, great! :ok_hand:

Jassob commented:
> I don't fully understands what tests are required for:
> 
> > I'm still missing test cases for checking if the bit pattern on the CAN frames are what we expected and if we can read/write messages that are compatible with other tools.

So what I'm after is some kind of check that our representation is correct according to some other tool. I want to be able to claim, with some certainty, that a float32 typed signal sent by us will be interpreted correctly by a third-party device or CAN tool.

Now that I'm thinking about it, we should be able to generate some kind of frame with cantool[^1], if it also supports floating point signals.

[^1]: https://cantools.readthedocs.io/en/latest/

Ezhik452 commented:
> Now that I'm thinking about it, we should be able to generate some kind of frame with cantool[1](#user-content-fn-1-b03f605b21212c54ae6da58bd969dee2), if it also supports floating point signals.
> ## Footnotes
> 
>     1. https://cantools.readthedocs.io/en/latest/ [↩](#user-content-fnref-1-b03f605b21212c54ae6da58bd969dee2)

Yes, the signals I'm using in marshal/unmarshal tests are produced by my node and verified with python cantools. 
I'm just a little confused with how this can be integrated into current golang tests.

Jassob commented:
> > Now that I'm thinking about it, we should be able to generate some kind of frame with cantool[1](#user-content-fn-1-b03f605b21212c54ae6da58bd969dee2), if it also supports floating point signals.
> > ## Footnotes
> > ```
> > 1. https://cantools.readthedocs.io/en/latest/ [↩](#user-content-fnref-1-b03f605b21212c54ae6da58bd969dee2)
> > ```
> 
> Yes, the signals I'm using in marshal/unmarshal tests are produced by my node and verified with python cantools. I'm just a little confused with how this can be integrated into current golang tests.

Right! I think the easiest way would be to manually create a `can.Frame` with the `can.Data` content being the bytes you see in candump. But then the unmarshal/marshal tests that you've already added tests that part already. I just didn't know the source of the frames. 

> For example, for signal of value `3.14`: ![Screenshot_20240523_110459](https://private-user-images.githubusercontent.com/64531926/333118441-6cef3bcd-eec7-4c27-a43d-f8237e601c7c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY0OTA3NTQsIm5iZiI6MTcxNjQ5MDQ1NCwicGF0aCI6Ii82NDUzMTkyNi8zMzMxMTg0NDEtNmNlZjNiY2QtZWVjNy00YzI3LWE0M2QtZjgyMzdlNjAxYzdjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTIzVDE4NTQxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMzZDNkMGU0OTY0YjNkYjU1MjY3OGQ5N2MwMGNiOGMyZjlmNzdhMmFkNWEyMzEzN2UzM2JkYTNiYmM4YzJjMDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.PcILoSlOG0uHp4r2ma4ivgUYQhooc7CFgsuXIcXmSbM)
> 
> On the left is `candump vcan0` and on the right `cantools monitor -c vcan0 my_node.dbc`

Awesome, I'll approve the PR :ok_hand:

Jassob commented:
@Ezhik452 The CI runs `make` as part of the GitHub Actions workflow, so make sure that you run this locally (make should download all tools needed to run the workflow) next time.

The issue this time around is that we run the dbc compiler on the directory `testdata/dbc`, so for that to not crash on our invalid files we can move the invalid DBC files to a new directory, such as `testdata/invalid-dbc`. Then we have to update the test cases that refer to those invalid DBC files as well.

Ezhik452 commented:
> Awesome, I'll approve the PR 👌

Thanks, awesome!

> The issue this time around is that we run the dbc compiler on the directory `testdata/dbc`, so for that to not crash on our invalid files we can move the invalid DBC files to a new directory, such as `testdata/invalid-dbc`. Then we have to update the test cases that refer to those invalid DBC files as well.

I moved files and fixed a couple of lines (`if` -> `switch` in `internal/generate/file.go`, removed type conversion to `uint64` in `pkg/descriptor/signal.go`) according to linter. Now `make` passes locally for me.

Jassob commented:
Thank you for your hard work and cooperation @Ezhik452!

Ezhik452 commented:
> Thank you for your hard work and cooperation @Ezhik452!

No problem. I'm glad to contribute something back to good projects :)
